partiql-eval/src/eval.rs

```diff
@@ -108,11 +108,103 @@ impl Evaluable for EvalScan {
         self.output.clone()
     }
 
-    fn update_input(&mut self, _input: &Value) {
+    fn update_input(&mut self, _input: &Value, _branch_num: u8) {
         todo!("update_input for Scan")
     }
 }
 
+#[derive(Debug)]
+pub enum EvalJoinKind {
+    Inner,
+    Left,
+    Right,
+    Full,
+    Cross,
+}
+
+#[derive(Debug)]
+pub struct EvalJoin {
+    pub kind: EvalJoinKind,
+    pub on: Option<Box<dyn EvalExpr>>,
+    pub input_l: Option<Value>,
+    pub input_r: Option<Value>,
+    pub output: Option<Value>,
+}
+
+impl EvalJoin {
+    pub fn new(kind: EvalJoinKind, on: Option<Box<dyn EvalExpr>>) -> Self {
+        EvalJoin {
+            kind,
+            on,
+            input_l: None,
+            input_r: None,
+            output: None,
+        }
+    }
+}
+
+impl Evaluable for EvalJoin {
+    fn evaluate(&mut self, ctx: &dyn EvalContext) -> Option<Value> {
```

Comment:
Currently, the `JOIN` implementation has quite a few `.clone()`s (partially from my own Rust inexperience). I'll work to remove unnecessary clones when we can agree on this modeling of `JOIN`s.